cl/validator/attestation_producer/attestation_producer.go

```diff
@@ -74,18 +80,45 @@ func (ap *attestationProducer) ProduceAndCacheAttestationData(baseState *state.C
 			baseAttestationData.Target(),
 		), nil
 	}
-	stateEpoch := state.Epoch(baseState)
+	ap.attCacheMutex.RUnlock()
 
-	if baseState.Slot() > slot {
-		return solid.AttestationData{}, errors.New("head state slot is bigger than requested slot, the attestation should have been cached, try again later.")
+	// in case the target epoch is not found, let's generate it with lock to avoid everyone trying to generate it
+	// at the same time, which would be a waste of memory resources
+	ap.attCacheMutex.Lock()
+	defer ap.attCacheMutex.Unlock()
+	// check again if the target epoch is already generated
+	if baseAttestationData, ok := ap.attestationsCache.Get(epoch); ok {
+		beaconBlockRoot := baseStateBlockRoot
+		if baseState.Slot() > slot {
+			beaconBlockRoot, err = baseState.GetBlockRootAtSlot(slot)
```

Comment:
I think this causes invalid attestations, can you check what the VC says? I think this is not 1 to 1 equivalent to what the non-cached case would return (talking specifically about `beaconBlockRoot`)